bitonic.hpp

```diff
@@ -0,0 +1,375 @@
+{
+    const std::size_t idx = blockDim.x * blockIdx.x + threadIdx.x;
+    // i = idx // 2^k * 2^(k+1) + idx % k where k=pass
+    // note idx//2^k * 2^(k+1) != 2*idx because integer division is floored
+    const std::size_t i = ((idx >> pass) << (pass + 1)) + ::fast_mod_pow_2(idx, pass);
+    const std::size_t j = ((idx >> pass) << (pass + 1)) + (1 << (pass + 1)) - ::fast_mod_pow_2(idx, pass) - 1;
+
+    if(/*j > i && */j < len && arr[j] < arr[i])
+    {
+        T tmp = arr[j];
+        arr[j] = arr[i];
+        arr[i] = tmp;
+    }
+}
+
+template<typename T>
+__global__ void bitonic_kernel_b(T* arr, std::size_t pass, std::size_t subpass, std::size_t len)
+{
+    const std::size_t idx = blockDim.x * blockIdx.x + threadIdx.x;
+    // i = idx // 2^k * 2^(k+1) + idx % k where k=(pass-subpass)
+    const std::size_t i = ((idx >> (pass - subpass - 1)) << (pass - subpass)) + ::fast_mod_pow_2(idx, pass - subpass - 1);
+    const std::size_t j = i + (1 << (pass - subpass - 1));
+
+    if(j < len && arr[j] < arr[i])
+    {
+        T tmp = arr[j];
+        arr[j] = arr[i];
+        arr[i] = tmp;
+    }
+}
+
+template<typename T>
+__host__ void bitonic_sort_gpu(T* h_arr, std::size_t len)
+{
+    const std::size_t fakelen = ::next_power_of_2(len);
+
+    T* d_arr;
+    const std::size_t arrsize = len * sizeof(T);
+    cudaMalloc(&d_arr, arrsize);
+
+    cudaMemcpy(d_arr, h_arr, arrsize, cudaMemcpyHostToDevice);
+
+    const std::size_t threadsPerBlock = 256; // does it matter?
+    // in principle we only need len/2 threads, not fakelen/2, but it would require some logic changes to skip some values of idx.
+    // Maybe a feature to add later.
+    const std::size_t n_blocks = (((fakelen + 1) / 2) + threadsPerBlock - 1) / threadsPerBlock;
+
+    for(std::size_t pass = 0; (1 << pass) < fakelen; ++pass)
+    {
+        bitonic_kernel_a<<<n_blocks, threadsPerBlock>>>(d_arr, pass, len);
+
+        for(std::size_t subpass = 0; subpass < pass; ++subpass)
+        {
+            bitonic_kernel_b<<<n_blocks, threadsPerBlock>>>(d_arr, pass, subpass, len);
+        }
+    }
+
+    cudaMemcpy(h_arr, d_arr, arrsize, cudaMemcpyDeviceToHost);
+
+    cudaFree(d_arr);
+}
+
+template<typename T>
+__device__ void bitonic_step_b_shared(T* shared, unsigned block_elems, unsigned pass, unsigned first_subpass, std::size_t len)
+{
+    // const unsigned block_elems = 1 << (pass - first_subpass);
+    const unsigned block_first_i = block_elems * blockIdx.x;
+    const unsigned thread_elems = (block_elems + blockDim.x - 1) / blockDim.x;
+
+    for(unsigned subpass = first_subpass; subpass < pass; ++subpass)
+    {
+        if(thread_elems >= (1 << (pass - subpass)))
+        {
+            for(unsigned k = 0; k < (thread_elems + 1) / 2; ++k)
+            {
+                // here, tier is within the elements this thread is responsible for
+                const unsigned tier = k / (1 << (pass - subpass - 1));
```

Comment:
I'm pretty sure the compiler will disappear these intermediate values, though I admit I'm not totally confident about the optimization differences between regular C++ code/compilers and CUDA kernels/NVCC/.